src/main/java/org/pitest/junit5/JUnit5TestUnitFinder.java

```diff
@@ -97,10 +102,15 @@ private class TestIdentifierListener implements TestExecutionListener {
         private final Class<?> testClass;
         private final TestUnitExecutionListener l;
         private final List<TestIdentifier> identifiers = synchronizedList(new ArrayList<>());
+        private final boolean serializeExecution;
+        private final Map<UniqueId, ReentrantLock> parentCoverageSerializers = new ConcurrentHashMap<>();

```

Comment:
We have a lot of locks here. I'd have imagined that we could force things to run serially using just a single lock. I'm sure that I'm wrong, but to understand what all these locks are doing I think a big comment block explaining what junit does and the rationale behind this.

Reply:
I imagined also a single lock, unfortunately at least for Spock it didn't work.
I added a wall of text explaining the situation and logic that hopefully makes it clearer.
I really was "a bit" silent in regards of comments here. :-D

Reply:
I also made the child locks lazily initialized atomic references now, so that the locks for childs are only created if actually needed like for data-driven Spock features.